ecosystem/indexer-grpc/indexer-grpc-utils/src/storage_format.rs

```diff
@@ -0,0 +1,789 @@
+            CacheEntry::GzipCompressionProto(bytes) => bytes.len(),
+            CacheEntry::Base64UncompressedProto(bytes) => bytes.len(),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TransactionsLegacyFile {
+    /// The version of the first transaction in the blob.
+    pub starting_version: u64,
+    /// The transactions in the blob.
+    #[serde(rename = "transactions")]
+    pub transactions_in_base64: Vec<String>,
+}
+
+/// FileStoreMetadata is the metadata for the file store.
+/// It's a JSON file with name: metadata.json.
+#[derive(Serialize, Deserialize, Copy, Clone, Debug, PartialEq, Eq)]
+pub struct FileStoreMetadata {
+    pub chain_id: u64,
+    // The size of each file folder, BLOB_STORAGE_SIZE, i.e., 1_000.
+    pub file_folder_size: usize,
+    // The current version of the file store.
+    pub version: u64,
+    // Storage format.
+    #[serde(default = "default_storage_format")]
+    pub storage_format: StorageFormat,
+}
+
+fn default_storage_format() -> StorageFormat {
+    StorageFormat::JsonBase64UncompressedProto
+}
+
+impl FileStoreMetadata {
+    pub fn new(chain_id: u64, version: u64, storage_format: StorageFormat) -> Self {
+        Self {
+            chain_id,
+            file_folder_size: FILE_ENTRY_TRANSACTION_COUNT as usize,
+            version,
+            storage_format,
+        }
+    }
+}
+
+pub struct CacheEntryBuilder {
+    // This is used to determine how to serialize the transaction.
+    // Do not use in the storage; format can be inferred or configured externally.
+    storage_format: StorageFormat,
+    transaction: Transaction,
+}
+
+impl CacheEntryBuilder {
+    pub fn new(transaction: Transaction, storage_format: StorageFormat) -> Self {
+        Self {
+            storage_format,
+            transaction,
+        }
+    }
+}
+
+pub struct CacheEntryKey {
+    pub version: u64,
+    pub storage_format: StorageFormat,
+}
+
+impl CacheEntryKey {
+    pub fn new(version: u64, storage_format: StorageFormat) -> Self {
+        Self {
+            version,
+            storage_format,
+        }
+    }
+}
+
+impl ToString for CacheEntryKey {
+    // Cacche key is generated based on the assumption that the naming prefix
+    // doesn't have impact on the cache key.
+    fn to_string(&self) -> String {
+        match self.storage_format {
+            StorageFormat::Bz2CompressedProto => {
+                format!("bz:{}", self.version)
+            },
+            StorageFormat::GzipCompressionProto => {
+                format!("gz:{}", self.version)
+            },
+            StorageFormat::UncompressedProto => {
+                format!("uc:{}", self.version)
+            },
+            // Legacy format.
+            StorageFormat::Base64UncompressedProto => {
+                format!("{}", self.version)
+            },
+            StorageFormat::JsonBase64UncompressedProto => {
+                format!("j64:{}", self.version)
+            },
```

Comment:
UncompressedProto, JsonBase64UncompressedProto types should never be in the cache. we should only handle the types we care about in these match arms